### PR TITLE
Added Reward Providers for Torch

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Adding the implementation of Curiosity and GAIL along with the reward provider.

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/reward_providers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from mlagents.trainers.reward_providers.base_reward_provider import (  # noqa F401
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/reward_providers/__init__.py:1]: Should all of these be put in a torch-specific sub-folder?

File: ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+import torch
+
+from mlagents.trainers.buffer import AgentBuffer
+from mlagents.trainers.reward_providers.base_reward_provider import BaseRewardProvider
+from mlagents.trainers.settings import CuriositySettings
+
+from mlagents_envs.base_env import BehaviorSpec
+
+
+def swish(x):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py:12]: We already have swish in the torch utils, I believe.

File: ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    for i in range(num_partitions):
+        res += [data[(partitions == i).nonzero().squeeze(1)]]
+    return res
+
+
+class CuriosityRewardProvider(BaseRewardProvider):
+    def __init__(self, specs: BehaviorSpec, settings: CuriositySettings) -> None:
+        super().__init__(specs, settings)
+        self._network = CuriosityNetwork(specs, settings)
+        params = list(self._network.parameters())
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py:88]: You can just pass `self._network.parameters()` directly to the optimizer. No need to turn it into a list.

File: ml-agents/mlagents/trainers/reward_providers/gail_reward_provider.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+        encoder_input_size = vec_obs_size  # + visuals later and normalization
+        if settings.use_actions and specs.is_action_continuous():
+            encoder_input_size += specs.action_size + 1  # + 1 is for done
+        if settings.use_actions and specs.is_action_discrete():
+            encoder_input_size += (
+                sum(specs.discrete_action_branches) + 1
+            )  # + 1 is for done
+
+        self.encoder = torch.nn.ModuleList(
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/reward_providers/gail_reward_provider.py:74]: You can use a `torch.nn.Sequential` here and avoid the for loop later on in the code.

File: ml-agents/mlagents/trainers/reward_providers/__init__.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/reward_providers/__init__.py:1]: Yes, I wanted to not have dependencies on other files for now, but these will definitely not stay here.

File: ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py (hunk 3 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py:88]: Nice, thanks

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 5]
```diff
@@ -86,18 +86,30 @@ def _process_trajectory(self, trajectory: Trajectory) -> None:
...
-            self._stats_reporter.add_stat(
-                self.optimizer.reward_signals[name].value_name, np.mean(v)
-            )
+            if hasattr(self.optimizer.reward_signals[name], "value_name"):
+                self._stats_reporter.add_stat(
+                    self.optimizer.reward_signals[name].value_name, np.mean(v)
+                )
+            else:
+                self._stats_reporter.add_stat(
+                    self.optimizer.reward_signals[name].name + "Value Estimate",
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/ppo/trainer.py:95]:
```suggestion
                    f"{self.optimizer.reward_signals[name].name} Value Estimate",
```

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 6]
```diff
@@ -86,18 +86,30 @@ def _process_trajectory(self, trajectory: Trajectory) -> None:
 
         for name, v in value_estimates.items():
             agent_buffer_trajectory[f"{name}_value_estimates"].extend(v)
-            self._stats_reporter.add_stat(
-                self.optimizer.reward_signals[name].value_name, np.mean(v)
-            )
+            if hasattr(self.optimizer.reward_signals[name], "value_name"):
+                self._stats_reporter.add_stat(
+                    self.optimizer.reward_signals[name].value_name, np.mean(v)
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/ppo/trainer.py:91]: Will not all reward signals have `value_name`? Should we just remove it completely and just use the `name`?

**Comment by ervteng** [comment]
Looks good to me!

File: ml-agents/mlagents/trainers/ppo/trainer.py (hunk 6 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/ppo/trainer.py:92]: That would work too

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 7]
```diff
@@ -86,18 +87,30 @@ def _process_trajectory(self, trajectory: Trajectory) -> None:
...
-            self._stats_reporter.add_stat(
-                self.optimizer.reward_signals[name].value_name, np.mean(v)
-            )
+            if isinstance(self.optimizer.reward_signals[name], RewardSignal):
+                self._stats_reporter.add_stat(
+                    self.optimizer.reward_signals[name].value_name, np.mean(v)
+                )
+            else:
+                self._stats_reporter.add_stat(
+                    f"Policy/{self.optimizer.reward_signals[name].name.capitalize()} Value Estimate",
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/ppo/trainer.py:96]: Seems weird to me. Can we always assume if it's not a `RewardSignal` it's a Value Estimate?

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 8]
```diff
@@ -86,18 +87,30 @@ def _process_trajectory(self, trajectory: Trajectory) -> None:
...
         # Evaluate all reward functions
         self.collected_rewards["environment"][agent_id] += np.sum(
             agent_buffer_trajectory["environment_rewards"]
         )
         for name, reward_signal in self.optimizer.reward_signals.items():
-            evaluate_result = reward_signal.evaluate_batch(
-                agent_buffer_trajectory
-            ).scaled_reward
+            if isinstance(reward_signal, RewardSignal):
+                evaluate_result = reward_signal.evaluate_batch(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/ppo/trainer.py:106]: Structurally, these two function calls look similar enough that maybe they can be aligned? Then we wouldn't need to check types.
